### PR TITLE
drivers/spi: Driver for the SPI peripheral on the SiFive FE310/HiFive 1

### PR DESCRIPTION
**Pull request by nategraff-sifive**
Adds a SPI driver for the SiFive FE310, or any other device which uses the SPI peripheral from [sifive-blocks](https://github.com/sifive/sifive-blocks) (Listed in the DTS as compatible with "sifive,spi0".)

This also makes the following configurations choices for the HiFive 1:
- The SPI 0 peripheral driver is not enabled by default because it is in charge of mapping the SPI flash into memory. This can be configured using the CONFIG_SIFIVE_SPI_0_ROM KConfig option.
- The SPI 1 peripheral driver is enabled by default and the pinmux is configured for all of its outputs
- The SPI 2 peripheral driver is enabled by default because it is present in the DTS for the FE310, but because the QFN48 package used on the HiFive 1 doesn't route those pins from the silicon die, the pinmux can't enable the SPI 2 pins.

When I rebased this driver on top of latest master, I got some errors on the shared spi_context.h code. (ex. LOG_DBG needed to be renamed SYS_LOG_DBG) If this is the wrong fix, let me know and I'll update my PR.

### CODE REVIEW

**Comment by nategraff-sifive** [comment]
Okay, so it looks like I found a catch-22: I can't build if I use LOG_*, and the CI fails a number of tests if I use SYS_LOG_*. I'll probably end up looking back into this after the holiday.

**Review by nashif — DISMISSED** [review]

**Comment by nategraff-sifive** [comment]
I'm seeing errors like the following when I build without moving back to the old system:
```
In file included from /scratch/nathanielg/zephyr/drivers/spi/spi_sifive.c:7:0:
/scratch/nathanielg/zephyr/drivers/spi/spi_context.h: In function ‘spi_context_cs_configure’:
/scratch/nathanielg/zephyr/drivers/spi/spi_context.h:177:3: warning: implicit declaration of function ‘LOG_INF’ [-Wimplicit-function-declaration]
   LOG_INF("CS control inhibited (no GPIO device)");
   ^~~~~~~
```

Is there maybe an issue with how I'm including the logging system in the driver? This is what I have now:
```
#define SYS_LOG_LEVEL CONFIG_SYS_LOG_SPI_LEVEL
#include <logging/sys_log.h>
```

File: dts/bindings/spi/spi.yaml [hunk 1]
```diff
@@ -19,6 +19,11 @@ properties:
       category: required
       description: compatible strings
       generation: define
+    clock-frequency:
+      type: int
+      category: optional
+      description: Clock frequency information for UART operation
```
Comment (nategraff-sifive) [inline dts/bindings/spi/spi.yaml]: Whoops, missed the description line when I copied this from the generic UART binding.

File: drivers/spi/spi_sifive.c [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+
+#include "spi_context.h"
+#include "spi_sifive.h"
+
+#include <stdbool.h>
+
+#include <device.h>
+#include <spi.h>
+
+#define MIN(x, y) ((x < y) ? x : y)
```
Comment (tbursztyka) [inline drivers/spi/spi_sifive.c]: No need for this, just use min() from libc

File: drivers/spi/spi_sifive.c [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+struct spi_sifive_cfg {
+	u32_t base;
+	u32_t f_sys;
+};
+
+/* Helper Functions */
+
+int spi_config(struct device *dev, u32_t frequency, u16_t operation)
+{
+	if (SPI_OP_MODE_GET(operation) != SPI_OP_MODE_MASTER)
```
Comment (tbursztyka) [inline drivers/spi/spi_sifive.c]: We add {} even for one-liners if/for/while/etc... in Zephyr. Please apply that everywhere.
(Note that you are doing it properly already in transceive function)

File: drivers/spi/spi_sifive.c [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+		 * on the trailing SCK edge
+		 */
+		SET_BIT(SPI_REG(dev, REG_SCKMODE), SF_SCKMODE_PHA);
+	} else {
+		/* Data is sampled on the leading SCK edge */
+		CLR_BIT(SPI_REG(dev, REG_SCKMODE), SF_SCKMODE_PHA);
+	}
+
+	/* Get the frame length */
+	u32_t fmt_len = SPI_WORD_SIZE_GET(operation);
```
Comment (tbursztyka) [inline drivers/spi/spi_sifive.c]: do not declare a variable in the middle of a code block, always at the beginning. Apply that everywhere

File: drivers/spi/spi_sifive.h [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+#define SET_BIT(reg, offset)    ((reg) |= (1 << (offset)))
+#define CLR_BIT(reg, offset)    ((reg) &= ~(1 << (offset)))
+#define CLR_MASK(reg, mask)     ((reg) &= ~(mask))
+
+#define SPI_CFG(dev) ((struct spi_sifive_cfg *) ((dev)->config->config_info))
+#define SPI_DATA(dev) ((struct spi_sifive_data *) ((dev)->driver_data))
+
+#define _REG32(p, i) (*(volatile u32_t *) ((p) + (i)))
+#define SPI_REG(dev, offset) _REG32(SPI_CFG(dev)->base, offset)
```
Comment (tbursztyka) [inline drivers/spi/spi_sifive.h]: seems like your are rewriting something that could be done with sys_io.h function. Change this to sys_<write/read>32(). (reason being if these is a better way - it's most probably very unlikely here though -  to access the memory, one change in sys_io implementation would benefit everywhere, and you would not have to do any change here)

File: drivers/spi/spi_sifive.c [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+
+struct spi_sifive_data {
+	struct spi_context ctx;
+};
+
+struct spi_sifive_cfg {
+	u32_t base;
+	u32_t f_sys;
+};
+
```
Comment (tbursztyka) [inline drivers/spi/spi_sifive.c]: Move these 2 structures to the header file.

File: drivers/spi/spi_sifive.c [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+	u32_t val;
+
+	while ((val = SPI_REG(dev, REG_RXDATA)) & SF_RXDATA_EMPTY)
+		;
+
+	return (u16_t) val;
+}
+
+int spi_sifive_transceive_one(struct device *dev,
+		const struct spi_buf *tx_buf,
```
Comment (tbursztyka) [inline drivers/spi/spi_sifive.c]: indentation of parameters is below last opened '(' + 1 space. Same for function calls, apply that everywhere (your editor should be able to automatically do that for you)

File: drivers/spi/spi_sifive.c [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+	const struct spi_buf *rx_buf = &empty_buf;
+
+	/* Perform transfers */
+	for (int i = 0; i < num_xfers; i++) {
+		if (tx_bufs != NULL)
+			tx_buf = &(tx_bufs->buffers[i]);
+		if (rx_bufs != NULL)
+			rx_buf = &(rx_bufs->buffers[i]);
+
+		rc = spi_sifive_transceive_one(dev,
```
Comment (tbursztyka) [inline drivers/spi/spi_sifive.c]: There is an issue here: all the given buffers in on spi_transceive() should be send/recv at once, i.e.: without releasing CS in between each. 

So I'd move the CS logic before/after this for loop (respectively for activating CS, disabling CS)

File: drivers/spi/spi_sifive.c [hunk 9]
```diff
@@ -0,0 +1,319 @@
...
+	/* Default to transferring dummy bytes/discarding received bytes */
+	const struct spi_buf empty_buf = {
+		.buf = NULL,
+		.len = 0,
+	};
+	const struct spi_buf *tx_buf = &empty_buf;
+	const struct spi_buf *rx_buf = &empty_buf;
+
+	/* Perform transfers */
+	for (int i = 0; i < num_xfers; i++) {
```
Comment (tbursztyka) [inline drivers/spi/spi_sifive.c]: Why not reusing spi_context helpers? so you would not have to have all this code around num_xfers, etc...

**Review by tbursztyka — DISMISSED** [review]

**Comment by nategraff-sifive** [comment]
Thanks @nashif and @tbursztyka! I just force-pushed all of the requested fixes.

I'm not 100% confident about how I used the spi_context helpers, since there doesn't seem to be much documentation surrounding their use. I'd appreciate a double-check, especially within `spi_sifive_xfer()`

**Comment by pyropeter** [comment]
The hardware chip select is used just like a GPIO (only set to OFF or HOLD), due to the uselessness of the hardware chip select logic in this chip. Wouldn't it simplify the code to remove the hardware chip select support entirely?

**Comment by AJGherardi** [comment]
When will this driver be available I need it for a project

**Comment by nategraff-sifive** [comment]
@AJGherardi Thanks for your interest. If you're dealing with something time-critical you could check out [the branch I'm merging from](https://github.com/sifive/riscv-zephyr/tree/sifive-spi-driver). You might consider rebasing on top of latest master if you do that.

@nashif Have your requested changes been addressed?

**Comment by AJGherardi** [comment]
Thank you I will do that

**Comment by kgugala** [comment]
looks OK

**Review by kgugala — APPROVED** [review]

**Review by tbursztyka — APPROVED** [review]

File: drivers/spi/Kconfig.sifive [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+# Copyright (c) 2018 SiFive Inc.
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+menuconfig SPI_SIFIVE
+	bool "SiFive SPI controller driver"
+	depends on SPI
+	depends on SOC_SERIES_RISCV32_SIFIVE_FREEDOM
+	depends on HAS_DTS_SPI
```
Comment (galak) [inline drivers/spi/Kconfig.sifive]: HAS_DTS_SPI is already done for all SPI controllers, so you don't need it.

**Review by galak — DISMISSED** [review]
Some minor dt comments,

Also can we use the following #defines in the driver so you don't need dts_fixup.h:

```
#define DT_SIFIVE_SPI0_0_BASE_ADDRESS_0                 DT_SIFIVE_SPI0_10014000_BASE_ADDRESS_0
#define DT_SIFIVE_SPI0_0_BASE_ADDRESS_1                 DT_SIFIVE_SPI0_10014000_BASE_ADDRESS_1
#define DT_SIFIVE_SPI0_0_CLOCK_FREQUENCY                DT_SIFIVE_SPI0_10014000_CLOCK_FREQUENCY
#define DT_SIFIVE_SPI0_0_CONTROL_BASE_ADDRESS           DT_SIFIVE_SPI0_10014000_BASE_ADDRESS_0
#define DT_SIFIVE_SPI0_0_CONTROL_SIZE                   DT_SIFIVE_SPI0_10014000_SIZE_0
#define DT_SIFIVE_SPI0_0_LABEL                          DT_SIFIVE_SPI0_10014000_LABEL
#define DT_SIFIVE_SPI0_0_MEM_BASE_ADDRESS               DT_SIFIVE_SPI0_10014000_BASE_ADDRESS_1
#define DT_SIFIVE_SPI0_0_MEM_SIZE                       DT_SIFIVE_SPI0_10014000_SIZE_1
#define DT_SIFIVE_SPI0_0_SIZE_0                         DT_SIFIVE_SPI0_10014000_SIZE_0
#define DT_SIFIVE_SPI0_0_SIZE_1                         DT_SIFIVE_SPI0_10014000_SIZE_1
```

**Comment by nategraff-sifive** [comment]
@galak Putting those #defines in the driver hard-codes the memory address of the SPI peripheral of a single SoC into the driver. Other SiFive SoCs can and do change the address of the SPI.

**Comment by galak** [comment]
> @galak Putting those #defines in the driver hard-codes the memory address of the SPI peripheral of a single SoC into the driver. Other SiFive SoCs can and do change the address of the SPI.

The defines I suggested are not based on address, but on a counting of the instance.  So from the driver perspective its just 0, 1, 2, 3, etc.. However if you need to know something specific about a given instance to setup pin muxing or such than we aren't quite there.

**Comment by nategraff-sifive** [comment]
@galak I figured it out after rebasing. I'll update the other driver PRs accordingly.

Thanks!

**Pull request merged**